release.sh

```diff
@@ -313,64 +314,109 @@ function build_from_source() {
 }
 
 function get_images_in_yamls() {
-  rm -rf imagerefs.txt
+  rm -rf "$IMAGES_REFS_FILE"
   echo "Assembling a list of image refences to sign"
-  for file in $@; do
+  for file in "$@"; do
     [[ "${file##*.}" != "yaml" ]] && continue
     echo "Inspecting ${file}"
-    for image in $(grep -oh "\S*${KO_DOCKER_REPO}\S*" "${file}"); do
-      echo $image >> imagerefs.txt
-    done
+    while read -r image; do
+      echo "$image" >> "$IMAGES_REFS_FILE"
+    done < <(grep -oh "\S*${KO_DOCKER_REPO}\S*" "${file}")
+  done
+  if [[ -f "$IMAGES_REFS_FILE" ]]; then
+    sort -uo "$IMAGES_REFS_FILE" "$IMAGES_REFS_FILE" # Remove duplicate entries
+  fi
+}
+
+function find_checksums_file() {
+  for file in "$@"; do
+    if [[ "${file}" == *"checksums.txt" ]]; then
+      echo "${file}"
+      return 0
+    fi
   done
-  sort -uo imagerefs.txt imagerefs.txt # Remove duplicate entries
+  warning "cannot find checksums file"
 }
 
 # Build a release from source.
 function sign_release() {
-  get_images_in_yamls "${ARTIFACTS_TO_PUBLISH}"
   if (( ! IS_PROW )); then # This function can't be run by devs on their laptops
     return 0
   fi
+  get_images_in_yamls "${ARTIFACTS_TO_PUBLISH}"
+  local checksums_file
+  checksums_file="$(find_checksums_file "${ARTIFACTS_TO_PUBLISH}")"
+
+  if ! [[ -f "${checksums_file}" ]]; then
+    echo '>> No checksums file found, generating one'
+    checksums_file="$(mktemp -d)/checksums.txt"
+    for file in ${ARTIFACTS_TO_PUBLISH}; do
+      pushd "$(dirname "$file")" >/dev/null
+      sha256sum "$(basename "$file")" >> "${checksums_file}"
+      popd >/dev/null
+    done
+    ARTIFACTS_TO_PUBLISH="${ARTIFACTS_TO_PUBLISH} ${checksums_file}"
+  fi
 
   # Notarizing mac binaries needs to be done before cosign as it changes the checksum values
   # of the darwin binaries
  if [ -n "${APPLE_CODESIGN_KEY}" ] && [ -n "${APPLE_CODESIGN_PASSWORD_FILE}" ] && [ -n "${APPLE_NOTARY_API_KEY}" ]; then
     banner "Notarizing macOS Binaries for the release"
-    FILES=$(find -- * -type f -name "*darwin*")
-    for file in $FILES; do
-      rcodesign sign "${file}" --p12-file="${APPLE_CODESIGN_KEY}" \
-        --code-signature-flags=runtime \
-        --p12-password-file="${APPLE_CODESIGN_PASSWORD_FILE}"
-    done
-    zip files.zip ${FILES}
-    rcodesign notary-submit files.zip --api-key-path="${APPLE_NOTARY_API_KEY}" --wait
-    sha256sum ${ARTIFACTS_TO_PUBLISH//checksums.txt/} > checksums.txt
-    echo "🧮     Post Notarization Checksum:"
-    cat checksums.txt
+    local macos_artifacts
+    declare -a macos_artifacts=()
```

Comment:
I don't see any glob-equal anywhere. Could you explain it exactly?

Reply:
On line 333, you wrote:

```
    if [[ "${file}" == *"checksums.txt" ]]; then
```

I wondered if that form would be better than `echo | grep`

Reply:
Oops, this was supposed to be on line 368:

```
      if echo "$file" | grep -q "darwin"; then
```

Reply:
Yes, it may be written as

```bash
if [[ "$file" == *"darwin"* ]]; then
```

IDK if it's worth changing though. This form is way less known, and it's very error prone.

For example, someone could rewrite it to:

```bash
if [[ "$file" == "*darwin*" ]]; then
```

Which means completely different thing, and will not work.

But, going back to your question: I'm not seeing much benefit in a rewrite, but we definitely could do it if you feel like it.